hv/crates/hv-cell/src/lib.rs

```diff
@@ -178,6 +179,75 @@ impl<T: ?Sized> ArcCell<T> {
     pub fn from_inner(inner: Arc<AtomicRefCell<T>>) -> Self {
         Self { inner }
     }
+
+    /// Construct a `WeakCell<T>`
+    pub fn downgrade(this: &Self) -> WeakCell<T> {
+        WeakCell {
+            inner: Arc::downgrade(&this.inner),
+        }
+    }
+}
+
+/// An analogue to `Arc<T>`'s `Weak<T>` type, normally constructed via `ArcCell::downgrade(this: &T)` or `WeakCell::new()`
+pub struct WeakCell<T: ?Sized> {
+    inner: Weak<AtomicRefCell<T>>,
+}
+
+impl<T> Default for WeakCell<T> {
+    fn default() -> Self {
+        WeakCell::new()
+    }
+}
+
+impl<T> WeakCell<T> {
+    #[inline]
+    /// Construct a new `WeakCell<T>`
+    pub fn new() -> Self {
+        WeakCell { inner: Weak::new() }
+    }
+
+    /// Retrieve the inner field of the `WeakCell<T>`
+    pub fn as_inner(&self) -> &Weak<AtomicRefCell<T>> {
+        &self.inner
+    }
+
+    /// Unsafely try to borrow the underlying `T` (note that this could panic either due to a `borrow`
+    /// error, or due to an `upgrade` error)
+    pub fn borrow(&self) -> ArcRef<T> {
+        self.upgrade().unwrap().borrow()
+    }
+
+    /// Unsafely try to mutably borrow the underlying `T` (note that this could panic either due to a `borrow_mut`
+    /// error, or due to an `upgrade` error)
+    pub fn borrow_mut(&mut self) -> ArcRefMut<T> {
+        self.upgrade().unwrap().borrow_mut()
+    }
+
+    /// Safely attempt to borrow the underlying `T`. It first attempts to upgrade the `WeakCell<T>` before
+    /// trying to borrow it.
+    pub fn try_borrow(&self) -> Result<ArcRef<T>, WeakBorrowError> {
+        self.upgrade()
+            .ok_or(WeakBorrowError::UpgradeError(UpgradeError { _private: () }))?
+            .try_borrow()
+            .map_err(WeakBorrowError::BorrowError)
+    }
+
+    /// Safely attempt to mutably borrow the underlying `T`. It first attempts to upgrade the `WeakCell<T>` before
+    /// trying to borrow it.
+    pub fn try_borrow_mut(&self) -> Result<ArcRefMut<T>, WeakBorrowMutError> {
+        self.upgrade()
+            .ok_or(WeakBorrowMutError::UpgradeError(UpgradeError {
+                _private: (),
+            }))?
+            .try_borrow_mut()
+            .map_err(WeakBorrowMutError::BorrowMutError)
+    }
+
+    /// Attempt to upgrade the `WeakCell<T>` into an `ArcCell<T>`. Returns `None` if the strong reference count for
```

Comment:
Strong/weak reference counts are an internal detail of `Arc`/`Weak`. This documentation should instead specify that it returns `None` if there are no longer any live strong references to the value, which is a clearer description of the semantics of the method.